### DISCUSSION TITLE
Fixing extends?

### DISCUSSION THREAD

**Discussion by louiss0** [nuxt/nuxt#16217] [Ideas]
As of right now, the `extends:` keyword allows people to access code from any repository on GitHub. Access to the code can come from anyone. I don't believe this should be the way it should work. Giving developers access to anyone's repository code is not a good idea. 

- If the code is a set of components a company's entire brand could be stolen because they decided to use components to structure icons or images. 
- When it comes to using composables or middleware someone could just steal that code and use it to do malicious stuff. 
- Even when it comes to something like an API the developer has access to that code. 

As far as I know, there are no restrictions for how extends works one of the most dangerous aspects of this is **pages**. 

If I have access to the pages that are created in another Nuxt repo then can't I just create clones of a website and present those pages as my own? 

What you have essentially done is allow developers to instantly take code from any Nuxt repo without giving someone a way to protect their code.

There is a way to fix this problem it's called a **restrict** config property. 

- When you pass in `true` it means that the user has full access if you pass in an array then the user doesn't have access to the folders specified.  `false` means full access

```
defineConfig({
 restrict: [
"pages",
"server"
]
})
```
 
Another thing that could be done is that the creator of a repository could be forced to define an `expose:` property. 
- `true`  means exposing all 
- `false` means exposing none
-  An array means to expose a set of folders

I personally don't think it's a good idea for someone to have access to the pages of another person's project or even their plugins. But those could be useful for microservices.

**Comment by danielroe** [comment]
I would love to know more about why you are concerned about remote sources.

I'm sure you realised, but for the avoidance of doubt, this feature does not 'give access' to repositories on GitHub, it can only clone public repositories (or private ones for which you have a valid access token), just the same as if you added them to your `package.json` by running `npm install --save https://github.com/my-org/my-repo`.

**Reply by louiss0** [reply]
I said the code, not the repo itself. I'll have to use it so that I can explain what I mean further. I saw it in a few videos. The point is that when it comes to extending GitHub repositories any developer can use the code that was written in every directory the project was written in. I'm not talking about the developer being able to change the code in the repository but just using it. 

To be clear I was talking about the developer just using the code from the repo. I realize using the word access is the wrong word here. It's using.   

**The point is that extends allows people to use code from other repositories without the writers of the other repositories being able to say no to some of the folders being used by the developer who wants to use the code.** 

If I'm correct you can extend pages and components. Being able to use `pages/` and `components/` from any other Nuxt project gives the developer the power to take a company's brand and site and present that as their own.

**Reply by danielroe** [reply]
Having a public repository on GitHub gives anyone the ability to clone that repository. The same kind of moral and ethical restrictions around passing someone else's design or code off as your own would still apply. So I'm not sure I understand the point you are making.

**Reply by Diizzayy** [reply]
@louiss0 Quite frankly, If the repository is publicly listed on Github, The owner is already well aware that anyone can come along and use their entire codebase without seeking permission.

**Reply by louiss0** [reply]
> Having a public repository on GitHub gives anyone the ability to clone that repository. The same kind of moral and ethical restrictions around passing someone else's design or code off as your own would still apply. So I'm not sure I understand the point you are making.

Ok then If you and most people don't have a problem with this then I will concede. So the way that I make it so that people can't just use the code from my repo is to make the GitHub repo private is that correct?

**Reply by TheAlexLichter** [reply]
> So the way that I make it so that people can't just use the code from my repo is to make the GitHub repo private is that correct?

Yes, that is correct. Always has been like that, independently of Nuxt 😋

**Reply by louiss0** [reply]
> > So the way that I make it so that people can't just use the code from my repo is to make the GitHub repo private is that correct?
> 
> Yes, that is correct. Always has been like that, independently of Nuxt 😋

I think it should be said in the docs then. I along with many people will not want people to use code from projects without their knowledge of this. I really think this feature will save lots of people. **But since you don't believe in my idea people should at least know how to keep their code protected from this feature.**

Thank you for your time.

**Reply by TheAlexLichter** [reply]
@louiss0 Just to clarify. Everything you publish on GitHub as public (non-private) can be used by anyone without your knowledge. This is not a Nuxt-related topic and the actual *intention* of GitHub - promoting open source

**Reply by tylerforesthauser** [reply]
Yeah this is like saying that there should be a warning on a map that features _public_ parks that this means ANYONE can go to these parks, even though that was clearly the case before this map was made. @louiss0

**Reply by louiss0** [reply]
> @louiss0 Just to clarify. Everything you publish on GitHub as public (non-private) can be used by anyone without your knowledge. This is not a Nuxt-related topic and the actual _intention_ of GitHub - promoting open source

I understand what you mean.  
I meant to say that people should be reminded that if they don't want their code to be used by the **Extends Feature** make it private. I concede based on the fact that I was paranoid

**Reply by IsraelOrtuno** [reply]
One more point here is that based on what I understand from what seems to be the upcoming documentation ([see the PR](https://github.com/nuxt/framework/pull/9405/files)) for the `extends` feature, that repository that you want to use to use must include a `nuxt.config.ts` file. Not every folder/repository would also match this requirement.

> minimal Nuxt layer directory should contain a `nuxt.config.ts` file to indicate it is a layer.
